### PR TITLE
Homework 4

### PR DESCRIPTION
**Pull request by EgorSmi**

### CODE REVIEW

File: include/afina/concurrency/Executor.h [hunk 1]
```diff
@@ -27,17 +29,41 @@ class Executor {
...
+    {
+        std::unique_lock<std::mutex> state_lock(mutex);
+        state = State::kStopping;
+        empty_condition.notify_all(); // to stop
+        if (await == true)
+        {
+            for (std::size_t i = 0; i<threads.size(); i++)
+                threads[i].join();
+        }
+        state = State::kStopped;
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: await = false => сразу выставляется Stopped - неверно. Stopped должен выставляться только при окончательной остановки threadpool (выставлять в последнем завершающемся потоке / здесь, если threadpool пустой)

File: include/afina/concurrency/Executor.h [hunk 2]
```diff
@@ -50,12 +76,19 @@ class Executor {
...
-        if (state != State::kRun) {
+        std::unique_lock<std::mutex> lock(mutex);
+        if (state != State::kRun || _cur_queue_size >= _max_queue_size) {
             return false;
         }
 
         // Enqueue new task
+        {
+            std::unique_lock<std::mutex> lock(pool_mutex);
+            if ((threads.size() < _high_watermark) && (_cur_queue_size++ >= _low_watermark))
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Отслеживать free_threads, создаем новый поток, если нет свободных

File: include/afina/concurrency/Executor.h [hunk 3]
```diff
@@ -27,17 +29,41 @@ class Executor {
         // Threadppol is stopped
         kStopped
     };
+public:
+    Executor(std::string name, std::size_t size, std::size_t low_watermark, std::size_t high_watermark,
+             std::size_t idle_time): state(State::kRun), _name(name), _max_queue_size(size), _low_watermark(low_watermark),
+                                     _high_watermark(high_watermark), _idle_time(idle_time), _cur_queue_size(0)
+    {
+        std::unique_lock<std::mutex> lock(pool_mutex);
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h:38]: Одного мьютекса достаточно

File: include/afina/concurrency/Executor.h [hunk 4]
```diff
@@ -71,7 +104,46 @@ class Executor {
     /**
      * Main function that all pool threads are running. It polls internal task queue and execute tasks
      */
-    friend void perform(Executor *executor);
+    friend void perform(Executor *executor)
+    {
+        std::unique_lock<std::mutex> lock(executor->mutex);
+        const std::chrono::time_point<std::chrono::system_clock> now = std::chrono::system_clock::now();
+        while (executor->state == State::kRun)
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: || !tasks.empty()

File: include/afina/concurrency/Executor.h [hunk 5]
```diff
@@ -27,17 +29,41 @@ class Executor {
...
-    void Stop(bool await = false);
+    void Stop(bool await = false)
+    {
+        std::unique_lock<std::mutex> state_lock(mutex);
+        state = State::kStopping;
+        empty_condition.notify_all(); // to stop
+        if (await == true)
+        {
+            for (std::size_t i = 0; i<threads.size(); i++)
+                threads[i].join();
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: 1) Потоки не смогут завершиться, т.к. мьютекс взят
2) Можем случайно сделать join() к завершенному потоку. Переписать на conditional_variable

File: include/afina/concurrency/Executor.h [hunk 6]
```diff
@@ -71,7 +104,46 @@ class Executor {
...
+                (executor->threads.size() > executor->_low_watermark))
+                 || (executor->state != State::kRun))
+                {
+                    if (executor->state == Executor::State::kRun)
+                    {
+                        std::unique_lock<std::mutex> pool_lock(executor->pool_mutex);
+                        auto thread_id = std::this_thread::get_id();
+                        auto it = std::find_if(executor->threads.begin(), executor->threads.end(),
+                                               [thread_id] (std::thread &t) { return t.get_id() == thread_id; });
+                        executor->threads.erase(it);
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h:134]: erase потока, который сейчас исполняется. Сначала detach()

File: include/afina/concurrency/Executor.h [hunk 7]
```diff
@@ -71,7 +104,46 @@ class Executor {
...
+        std::unique_lock<std::mutex> lock(executor->mutex);
+        const std::chrono::time_point<std::chrono::system_clock> now = std::chrono::system_clock::now();
+        while (executor->state == State::kRun)
+        {
+            std::function<void()> task; // task to perform
+            while (executor->tasks.empty())
+            {
+                // wait cv
+                auto cv_flag = executor->empty_condition.wait_until(lock,
+                                                                    now + std::chrono::milliseconds(executor->_idle_time));
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h:122]: now обновлять каждый раз

File: include/afina/concurrency/Executor.h [hunk 8]
```diff
@@ -71,7 +104,46 @@ class Executor {
...
+                    }
+                    return;
+                }
+            }
+            task = std::move(executor->tasks.front()); // take task from queue
+            executor->_cur_queue_size--;
+            executor->tasks.pop_front(); // destroy first element
+            try
+            {
+                task();
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h:155]: Таска исполняется не под мьютексом

File: include/afina/concurrency/Executor.h [hunk 9]
```diff
@@ -71,7 +104,46 @@ class Executor {
...
+            }
+            catch(...)
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: catch(const std::exception&) - если можно что-то залоггировать

**Review by Fen99 — CHANGES_REQUESTED** [review]

File: include/afina/concurrency/Executor.h [hunk 10]
```diff
@@ -27,17 +30,45 @@ class Executor {
...
      * free. All enqueued jobs will be complete.
      *
      * In case if await flag is true, call won't return until all background jobs are done and all threads are stopped
      */
-    void Stop(bool await = false);
+    void Stop(bool await = false)
+    {
+        std::unique_lock<std::mutex> state_lock(mutex);
+        state = State::kStopping;
+        state_lock.unlock();
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Не нужно

File: include/afina/concurrency/Executor.h [hunk 11]
```diff
@@ -50,12 +81,16 @@ class Executor {
...
 
-        std::unique_lock<std::mutex> lock(this->mutex);
-        if (state != State::kRun) {
+        std::unique_lock<std::mutex> lock(mutex);
+        if (state != State::kRun || _cur_queue_size >= _max_queue_size) {
             return false;
         }
 
         // Enqueue new task
+        if ((threads.size() < _high_watermark) && (_cur_queue_size++ >= 0))
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Запускаем новый, если нет свободных потоков

File: include/afina/concurrency/Executor.h [hunk 12]
```diff
@@ -71,7 +106,69 @@ class Executor {
...
+                    else
+                    {
+                        goto_delete = true;
+                        break;
+                    }
+                }
+            }
+            if (goto_delete)
+            {
+                break;
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: continue;

File: include/afina/concurrency/Executor.h [hunk 13]
```diff
@@ -71,7 +106,69 @@ class Executor {
...
+        }
+        // Stopping thread pool
+        auto thread_id = std::this_thread::get_id();
+        auto it = std::find_if(executor->threads.begin(), executor->threads.end(),
+                               [thread_id] (std::thread &t) { return t.get_id() == thread_id; });
+        (*it).detach();
+        executor->threads.erase(it);
+        if (executor->threads.empty())
+        {
+            executor->stop_condition.notify_all();
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h:173]: kStopped установить, если это последний

**Review by Fen99 — CHANGES_REQUESTED** [review]

File: include/afina/concurrency/Executor.h [hunk 14]
```diff
@@ -27,17 +30,43 @@ class Executor {
...
      *
      * In case if await flag is true, call won't return until all background jobs are done and all threads are stopped
      */
-    void Stop(bool await = false);
+    void Stop(bool await = false)
+    {
+        std::unique_lock<std::mutex> state_lock(mutex);
+        state = State::kStopping;
+        empty_condition.notify_all(); // to stop
+        if (await == true)
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: && !threads.empty()

**Review by Fen99 — APPROVED** [review]
С учетом небольшого исправления
